### PR TITLE
add magefiles

### PR DESCRIPTION
**Pull request by vroldanbet**
- introduces mage to handle common tasks
- improves proto generation so all tooling is pinned and it's run in CI exactly as it is locally
- updates all GH actions to use mage

### CODE REVIEW

File: buf.gen.yaml [hunk 1]
```diff
@@ -1,23 +1,29 @@
-#!/usr/bin/env -S buf generate buf.build/authzed/api:125e39e67bf60b2d19c4de15aaa69cb1bcbda28c --template
+#!/usr/bin/env -S go run github.com/bufbuild/buf/cmd/buf generate buf.build/authzed/api:125e39e67bf60b2d19c4de15aaa69cb1bcbda28c --template
```
Comment (ecordell) [inline buf.gen.yaml:1]: this makes me think there's something we could do (in the future) with the buf api. fetch the latest tag and generate off it or something.

File: magefile.go
Comment (ecordell) [inline magefile.go:1]: for consistency, name this `mage.go`?

File: .github/workflows/lint.yaml [hunk 2]
```diff
@@ -16,19 +16,33 @@ jobs:
...
-      - uses: "actions/setup-go@v4"
+      - uses: "authzed/actions/setup-go@main"
         with:
-          go-version: "~1.19"
-      - uses: "authzed/actions/gofumpt@main"
-      - uses: "authzed/actions/go-mod-tidy@main"
-      - uses: "authzed/actions/go-generate@main"
-      - uses: "authzed/actions/golangci-lint@main"
+          go-version: "${{ env.GO_VERSION }}"
+      - name: "Lint Go"
```
Comment (ecordell) [inline .github/workflows/lint.yaml:22]: Shouldn't this have a `deps:tidy` step in it to replace the action?

File: magefile.go
Comment (vroldanbet) [reply magefile.go:1]: actually when you do `mage -init` it generates a `magefile.go`, why deviating from what the tool proposes?

File: .github/workflows/lint.yaml (hunk 2 above)
Comment (vroldanbet) [reply .github/workflows/lint.yaml:22]: ah good catch

File: buf.gen.yaml (hunk 1 above)
Comment (vroldanbet) [reply buf.gen.yaml:1]: that'd be nice, something that updates the tag used, and then generates the code

File: magefile.go
Comment (ecordell) [reply magefile.go:1]: That's frustrating; I just based it off of the docs: https://magefile.org/zeroinstall/

**Pull request merged**